File: DFS & BFS/boj11724.py
import sys
sys.setrecursionlimit(10**6)

def dfs(graph, v, visited):
    if visited[v]:
        return False
    
    visited[v] = True
    for neighbor in graph[v]:
        if not visited[neighbor]:
            dfs(graph, neighbor, visited)
            
    return True

input = sys.stdin.readline

n, m = map(int, input().split())

graph = [[] for _ in range(n+1)]
visited = [False] * (n+1)
for _ in range(m):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)
    
count = 0
for i in range(1, n+1):
    if dfs(graph, i, visited):
        count += 1
        
print(count)

File: DFS & BFS/Boj11724.java
import java.util.ArrayList;
import java.util.Scanner;

public class Boj11724 {
    public static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    public static boolean[] visited;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int m = sc.nextInt();

        for(int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }
        visited = new boolean[n+1];

        for(int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        int count = 0;
        for(int i = 1; i <= n; i++) {
            if(!visited[i]) {
                dfs(i);
                count++;
            }
        }
        System.out.println(count);
    }

    public static void dfs(int x) {
        visited[x] = true;

        for(int neighbor : graph.get(x)) {
            if(!visited[neighbor]) {
                dfs(neighbor);
            }
        }
    }
}
